docs/encyclopedia/application-message-passing.mdx

```diff
@@ -220,22 +220,40 @@ Every time the Workflow wakes up--generally, it wakes up when it needs to--it wi
 
 This execution is on a single thread–while this means you don’t have to worry about parallelism, you do need to worry about concurrency if you have written Signal and Update handlers that can block. These can run interleaved with the main Workflow and with one another, resulting in potential race conditions. These methods should be made reentrant.
 
+#### Initializing the Workflow first {#workflow-initializers}
+
+Initialize your Workflow's state before handling messages. This prevents reading uninitialized instance variables.
+
+To see why, refer to the [diagram](#message-handler-concurrency). It shows that your Workflow processes messages before the first run of your Workflow's main method.
+
+This happens in several scenarios, such as:
+
+- When using [Signal-with-Start](#signal-with-start), the message handler always runs first.
+- When your Worker experiences delays, such as during backlogs.
+- When messages arrive immediately after a Workflow continues as new but before it resumes.
+
+For all languages except Go and TypeScript, use your constructor to set up state. Annotate your constructor as an Workflow Initializer and take the same arguments as your Workflow's main method.
+
+Note that you can't make blocking calls from your constructor. If you need to block, make your Signal or Update handler [wait](#waiting) for an initialization flag.
+
+In Go and TypeScript, register any message handlers only after completing initialization.
+
 ### Message handler patterns {#message-handler-patterns}
 
 Here are several common patterns for write operations, Signal and Update handlers. They don't apply to pure read operations, i.e. Queries or [Update Validators](#update-validators):
 
-- Synchronous, returning immediately.
-- Waiting for the Workflow to be ready to process them.
+- Returning immediately from a handler.
+- Waiting for the Workflow to be ready to process them
 - Kicking off activities and other asynchronous tasks
-- Injecting work into the main Workflow.
-- Finishing handlers before the Workflow completes.
-- Ensuring your messages are processed exactly once.
+- Injecting work into the main Workflow
+- Finishing handlers before the Workflow completes
```

Comment:
Only because of previous tickets do I know that this really means  something along the lines of blocking the Workflow's completion until all handlers have finished processing, otherwise I don't know if it really would be meaningful to me at first read.

Suggest minor rewording to make it clearer

Reply:
Discussed in Slack. Drew prefers to leave as is.